crates/driver/src/settlement_proposal.rs

```diff
@@ -0,0 +1,356 @@
+                encoder.add_liquidity_order_trade(
+                    trade.order,
+                    trade.executed_amount,
+                    remaining_amounts.fee_amount,
+                )?;
+            } else {
+                encoder.add_trade(
+                    trade.order,
+                    trade.executed_amount,
+                    remaining_amounts.fee_amount,
+                )?;
+            }
+        }
+
+        let futures = self
+            .execution_plan
+            .iter()
+            .map(|interaction| interaction.finalize());
+        let interactions = futures::future::try_join_all(futures).await?;
+        interactions
+            .into_iter()
+            .flat_map(|i| i.encode())
+            .for_each(|i| encoder.append_to_execution_plan(i));
+
+        Ok(encoder)
+    }
+
+    pub async fn into_settlement(self) -> Result<Settlement> {
+        Ok(Settlement {
+            encoder: self.into_encoder().await?,
+        })
+    }
+
+    /// Computes the `SettlementSummary` if following checks are successful:
+    ///   - individual trades don't violate required properties
+    ///   - enough token balances before each on-chain interaction
+    ///   - enough token balances to pay out orders at the end
+    ///   - token conservation (TODO)
+    pub fn into_settlement_summary(
+        &self,
+        gas_price: f64,
+        external_prices: &ExternalPrices,
+    ) -> Result<SettlementSummary> {
+        let mut balances = HashMap::<H160, U256>::default();
+        let mut gas_used = U256::zero();
+
+        let mut trade_executions = Vec::with_capacity(self.trades.len());
+        for trade in &self.trades {
+            trade.sanity_check_trade_properties(&self.clearing_prices)?;
+            let execution = trade
+                .execution(&self.clearing_prices)
+                .with_context(|| format!("could not compute trade execution: {trade:?}"))?;
+            trade_executions.push(execution);
+        }
+
+        for (trade, execution) in self.trades.iter().zip(&trade_executions) {
+            let balance = balances.entry(execution.sell_token).or_default();
+            *balance = balance
+                .checked_add(execution.sell_amount)
+                .with_context(|| format!("order would overflow balance: {trade:?}"))?;
+        }
+
+        for interaction in &self.execution_plan {
+            let meta = interaction.metadata();
+            for input in &meta.inputs {
+                match balances.entry(input.token) {
+                    Entry::Occupied(mut entry) => {
+                        *entry.get_mut() =
+                            entry.get().checked_sub(input.amount).with_context(|| {
+                                format!("interaction would underflow balance: {:?}", &meta)
+                            })?;
+                    }
+                    _ => anyhow::bail!(format!("no balance for interaction: {:?}", input.token)),
```

Comment:
I also think its fine for balances to go negative within a settlement. A smart solver can take an "atomic" loan on settlement contract balances to get a better settlement.

Reply:
Simply allowing any balance to underflow is too permissive IMO. A simulation should fail if the solution "loans" more than the settlement contract currently holds or if it does internal buffer trades with untrusted tokens.
Maybe we should initialize the balances with the current balance of the settlement contract and also assert that the final contract balances have to be >= than at the start (unless internal buffer trades are allowed)?

Reply:
> Simply allowing any balance to underflow is too permissive IMO. A simulation should fail if the solution "loans" more than the settlement contract currently holds

Right, so in my mind, we would allow any amount of underflow, and if it were more than the maximum (i.e. how much the settlement holds in its buffers) then the settlement simulation that we do later would fail. That being said, if we want to check this at the commit stage of the solving process (i.e. before we have calldata), then I agree that we should initialize balances to the current buffer balances and then check at the end that all balances >= buffer balances.

> unless internal buffer trades are allowed

I think we in the future, we want to provide interaction calldata for internalized buffer trades. With this in mind, we would still do the check with the interaction that is meant to be internalized and just execute the settlement with that interaction omitted. So I don't think we need a "special" case for this here.

Reply:
Ideally I would like to design a solution which allows solvers to give the entire call data in the commit stage. If a solver does that we should still just do a "real" simulation to see if the solution is viable.
However, I think that RFQ support will have a huge impact on the competitiveness of solvers so it would make sense for many solutions to not be properly simulatable (missing call data). Because the commit-reveal approach of the new driver design handles it quite poorly when a winning solution does not work in the end I would like to do as many checks as we can during the "fake" simulation to avoid that problem as much as possible.

Reply:
Since the simulation also includes adding the `fee_amounts` of each order the solution actually is not allow to let the final balance fall below `initial_balances + all_fee_amounts`, right? Otherwise the solution would have traded away the protocol's fees.

Reply:
> Otherwise the solution would have traded away the protocol's fees.

It is technically allowed to do so (although it will get charged for this during solver payouts). This is what I meant by "bribing" - it gives fees back to the users to increase surplus and win, the cost to the solver is that it will get this amount discounted during the solver payout, but they can still be profitable this way (for example, in the case where an order's fee amount is much higher than what is actually needed to execute the trade on-chain, a cleaver solver can give back some of these fees to the user to improve their objective value in order to win the settlement and still stay in the green).

Reply:
This is a delicate point, and we had to remove the global token conservation constraint, as we call it from it, as people were not happy with it; see bribing/pennying, or simply a prediction that there will be a price shift. As we are, in any case, doing the accounting at the end of the week, I don't think we should do this check, unless we make it a very liberal one (as in, the balance of a token cannot be reduced by, say an equivalent of $500 or something like that).

Reply:
> I don't think we should do this check, unless we make it a very liberal one

We can still keep the check that ensures we don't go into negative balances (because that would cause reverts).